Find derivative recurrences from PDE (via parametric derivative taking)

hirish99 opened the pull request:
Revised version of https://github.com/inducer/sumpy/pull/202, using parametric approach to compute recurrences for derivatives.

hirish99 commented:
Requesting review @inducer.

hirish99 commented:
Requesting review @inducer

hirish99 commented:
Request review @inducer

hirish99 commented:
Request review

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+
+    :arg recurrence: a recurrence relation in :math:`s(n)`
+    """
+    terms = list(r.atoms(sp.Function))
```
```suggestion
    # We're assuming here that s(...) are the only function calls.
    terms = list(r.atoms(sp.Function))
```

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    # Sort terms and create idx_l
+    idx_l = []
+    for i in range(len(terms)):
+        tms = list(terms[i].atoms(sp.Number))
```
Carefully destructure `term` for correctness.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    :arg recurrence: a recurrence relation in :math:`s(n)`
+    """
+    terms = list(r.atoms(sp.Function))
+    terms = np.array(terms)
```
No reason for numpy arrays here.

hirish99 replied:
Used to sort terms later.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+
+    # Sort terms and create idx_l
+    idx_l = []
+    for i in range(len(terms)):
```
Comprehension

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+                           for i in range(0, len(terms)-1)])
+    true_recurrence1 = true_recurrence.subs(n, n-shift_idx)
+
+    if replace:
+        # Replace s(n-1) with snm_1, s(n-2) with snm_2 etc.
+        # because pymbolic.substitute won't recognize it
+        last_syms = [sp.Symbol(f"anm{i+1}") for i in range(order-1)]
+        # pylint: disable=not-callable
+        # Assumes order > 1
+        true_recurrence2 = true_recurrence1.subs(s(n-1), last_syms[0])
+        for i in range(2, order):
+            true_recurrence2 = true_recurrence2.subs(s(n-i), last_syms[i-1])
+        return order, true_recurrence2
```
Unused? Kill it.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    idx_l = idx_l[idx_sort]
+    terms = terms[idx_sort]
+
+    # Order is the max difference between highest/lowest in idx_l
+    order = max(idx_l) - min(idx_l) + 1
+
+    # How much do we need to shift the recurrence relation
+    shift_idx = max(idx_l)
+
+    # Get the respective coefficients in the recurrence relation from r
+    n = sp.symbols("n")
+    s = sp.Function("s")
+    coeffs = sp.poly(r, list(terms)).coeffs()
+
+    # Re-arrange the recurrence relation so we get s(n) = ____
+    # in terms of s(n-1), ...
+    true_recurrence = sum([coeffs[i]/coeffs[-1] * terms[i]
+                           for i in range(0, len(terms)-1)])
+    true_recurrence1 = true_recurrence.subs(n, n-shift_idx)
```
Why not solve for the highest-order "s"?

hirish99 replied:
Good suggestion, but after looking at this I would not want the denominators to collect after calling solve since I need to perform a re-index later. I want to keep the s(i) terms separate.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    return np.any(idx_l < 0)
+
+
+def __get_initial_c(recurrence):
```
```suggestion
def _get_initial_c(recurrence):
```

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+
+
+def __get_initial_c(recurrence):
+    r"""
+    For a given recurrence checks how many initial conditions by
+    checking for non-negative indexed terms.
+    """
```
Improve, add example?

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    return i
+
+
+def shift_recurrence(r: sp.Expr, var: np.ndarray) -> sp.Expr:
```
Can we generate those in the first place, to avoid having to have two "colors" of recurrence?

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    interactions_2d = 0
+    for i in range(p+1):
+        lamb_expr = generate_lamb_expr(i, n_initial)
+        a = storage[-4:] + [cts_r_s[0],cts_r_s[1],radius]
```
-4?

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+        exp_res = recurrence_qbx_lp(sources, centers, normals, strengths, radius, laplace2d, g_x_y, p)
+        qbx_res = qbx_lp_laplace_general(sources, sources, centers, radius, strengths, p)
+        #qbx_res,_ = lpot_eval_circle(sources.shape[1], p)
+        err.append(np.max(exp_res - qbx_res))
```
Add an assert on the errors.

inducer replied:
Use relative error.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    return exp_res
+
+
+# TEST CODE
```
Remove this.

inducer commented on test/test_recurrence.py:
```diff
@@ -0,0 +1,83 @@
+from sumpy.expansion.diff_op import (
```
Add license header.

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,607 @@
+    return cts_rotated_shifted
+
+
+# ================ Recurrence LP Eval =================
```
Separate file.

inducer commented on test/test_recurrenceqbx.py:
```diff
@@ -0,0 +1,151 @@
+    return result_qbx
+
+
+def _qbx_lp_laplace_general(sources, targets, centers, radius, strengths, order):
```
Unify with the Helmholtz case?

inducer commented on test/test_recurrenceqbx.py:
```diff
@@ -0,0 +1,151 @@
+    return sources, centers, normals, density, h, radius
+
+
+def test_recurrence_laplace_2d_ellipse():
```
Unify with Helmholtz case?

inducer commented on test/test_recurrence.py:
```diff
@@ -0,0 +1,201 @@
+                      var_t[0], i).subs(var_t[0], 0).subs(var_t[1], 0).subs(var_t[2], 0)
+                                               for i in range(6)]
+
+    # pylint: disable-next=not-callable
+    subs_dict = {s(0): derivs[0], s(1): derivs[1]}
+    check_2_s = r.subs(n, 2).subs(subs_dict) - derivs[2]
+    # pylint: disable-next=not-callable
+    subs_dict[s(2)] = derivs[2]
+    check_3_s = r.subs(n, 3).subs(subs_dict) - derivs[3]
+    # pylint: disable-next=not-callable
+    subs_dict[s(3)] = derivs[3]
+    check_4_s = r.subs(n, 4).subs(subs_dict) - derivs[4]
+    # pylint: disable-next=not-callable
+    subs_dict[s(4)] = derivs[4]
+    check_5_s = r.subs(n, 5).subs(subs_dict) - derivs[5]
+
+    x_coord = np.random.rand()  # noqa: NPY002
+    y_coord = np.random.rand()  # noqa: NPY002
+    z_coord = np.random.rand()  # noqa: NPY002
+    coord_dict = {var[0]: x_coord, var[1]: y_coord, var[2]: z_coord}
+
+    assert abs(check_2_s.subs(coord_dict)) <= 1e-15
+    assert abs(check_3_s.subs(coord_dict)) <= 1e-14
+    assert abs(check_4_s.subs(coord_dict)) <= 1e-12
+    assert abs(check_5_s.subs(coord_dict)) <= 1e-12
```
Make this a loop (also below).

inducer commented on test/test_recurrence.py:
```diff
@@ -0,0 +1,201 @@
+    assert abs(abs(check_3_s.subs(coord_dict))) <= 1e-14
+    assert abs(abs(check_4_s.subs(coord_dict))) <= 1e-12
+    assert abs(abs(check_5_s.subs(coord_dict))) <= 1e-12
+
+
+test_laplace2d()
+test_helmholtz2d()
+test_helmholtz3d()
+test_laplace3d()
```
```suggestion
```

inducer commented on test/test_recurrenceqbx.py:
```diff
@@ -0,0 +1,151 @@
+        #qbx_res,_ = lpot_eval_circle(sources.shape[1], p)
+        err.append(np.max(np.abs(exp_res - qbx_res)))
+    assert np.max(err) <= 1e-13
+
+test_recurrence_helmholtz_2d_ellipse()
```
```suggestion
```

inducer commented on test/test_recurrence.py:
```diff
@@ -0,0 +1,201 @@
+    x_coord = np.random.rand()  # noqa: NPY002
+    y_coord = np.random.rand()  # noqa: NPY002
+    coord_dict = {var[0]: x_coord, var[1]: y_coord}
+    derivs = [derivs[i].subs(coord_dict) for i in range(6)]
```
```suggestion
    derivs = [d.subs(coord_dict) for d in derivs]
```

inducer commented on test/test_recurrence.py:
```diff
@@ -0,0 +1,201 @@
+    r"""
+    Tests recurrence code for orders up to 6 helmholtz2d.
+    """
+    w = make_identity_diff_op(2)
+    helmholtz2d = laplacian(w) + w
+    _, _, r = get_processed_and_shifted_recurrence(helmholtz2d)
+
+    n = sp.symbols("n")
+    s = sp.Function("s")
+
+    var = _make_sympy_vec("x", 2)
+    var_t = _make_sympy_vec("t", 2)
+    abs_dist = sp.sqrt((var[0]-var_t[0])**2 +
+                       (var[1]-var_t[1])**2)
+    k = 1
+    g_x_y = (1j/4) * hankel1(0, k * abs_dist)
```
Get from a `Kernel` object? (Here and elsewhere.)

inducer commented on sumpy/recurrence.py:
```diff
@@ -0,0 +1,406 @@
+    return i
+
+
+def shift_recurrence(r: sp.Expr) -> sp.Expr:
+    r"""
+    A function that "shifts" the recurrence so it's center is placed
+    at the origin and source is the input for the recurrence generated.
+
+    :arg recurrence: a recurrence relation in :math:`s(n)`
+    """
```
Explain better.

hirish99 commented:
Plots updated in plot_taylor_recurrence and plot_normal_recurrence

hirish99 committed:
Add parametric recurrence QBX implementation

Adds recurrence-based QBX layer potential evaluation with on-axis
and off-axis computation support for Laplace and Helmholtz kernels
in 2D and 3D.

hirish99 committed:
Fix basedpyright errors, remove plotting code, update gitignore

- Add pyright ignore comments for sympy typing issues in recurrence.py
  and recurrence_qbx.py (sum() return type, .subs() return type,
  ObjectArray1D vs ndarray)
- Broaden parameter types from sp.Expr to sp.Basic where .subs() results
  are passed
- Remove personal plotting code from test files
- Remove tracked experimental files (pgf, png, svg, ipynb)
- Update .gitignore to exclude *.ipynb, *.pgf, output.png, qbxrecurrence.svg

hirish99 committed:
Fix basedpyright errors in test files and documentation build issues

- Add pyright ignore comments for reportMissingImports, reportCallIssue,
  reportArgumentType, reportAttributeAccessIssue, and reportIndexIssue
  in test files (meshmode/pytential are optional deps without full stubs)
- Fix RST formatting in recurrence.py docstrings (indentation, bullet lists)
- Fix wrong function name in autofunction directive
- Fix typo in class reference (LinearSystemPDEOperator -> LinearPDESystemOperator)

hirish99 committed:
Move pyright ignore to from-line for parenthesized imports

basedpyright reports reportMissingImports on the 'from' line,
not on the inner item line where ruff moved the comment.

hirish99 committed:
Rename on-axis/off-axis to large/small-|x_1| terminology, add docs

Rename functions to match paper (arXiv:2509.03687) terminology:
- get_on_axis_recurrence -> get_large_x1_recurrence
- get_off_axis_recurrence -> get_small_x1_recurrence
- get_off_axis_expression -> get_small_x1_expansion

Add documentation with diagram and usage examples for computing
derivatives in both regimes. Add recurrence_qbx to docs.

hirish99 committed:
Improve recurrence module documentation with LaTeX equations and clearer notation

hirish99 committed:
Update basedpyright baseline

hirish99 committed:
Fix docstring and merge basedpyright baseline with upstream main

- Fix recurrence_qbx_lp docstring: evaluation is at target points
  (center + radius * normal), not at expansion centers.
- Merge baseline with upstream/main to cover CI-specific diagnostics
  for pre-existing files while adding entries for new recurrence files.

hirish99 commented:
Request review.

1. Changed function names to reflect updates to paper.
2. Made documentation clear with diagram/examples on how to use recurrence code to compute derivatives.
3. Made all code compute derivative at target centered at 0, with arbitrary source, only convention allowed. Removed unneeded "shifting" code.

Pyright issues are due sympy code, cannot fix.

hirish99 committed:
Merge branch 'main' into parametric_recurrence

hirish99 committed:
recurrence: route f(rval) chain rule through an AppliedUndef placeholder

Modern SymPy's chain rule on f(rval) with rval = sqrt(sum(x**2)) + eps
routes through _diff_wrt(rval), which now strictly requires a Symbol-like
target and rejects sums, breaking every .diff(...) on f(rval) with
"Can't calculate derivative wrt epsilon + sqrt(...)". Replace rval inside
f(...) with R = Function("R")(*var, eps); the chain rule is happy with
AppliedUndef, then fold R and its abstract derivatives back to rval
after the f_r identification step. Use xreplace (not subs) so the
identification doesn't recurse into Derivative interiors. Substitute
eps-derivatives of R high-to-low + doit() to avoid leftover
Derivative(1, eps) residue. The parametric-eps recurrence algorithm
itself is unchanged; only the SymPy carrier of rval changes.

Also drop a stale `# pyright: ignore[reportCallIssue]` on
test_recurrence_qbx.py:142 and switch the on/off-axis blend threshold in
recurrence_qbx_lp from m=100 to xi=10.

hirish99 committed:
recurrence: fix ruff (N802/N803/N806/PIE808) and basedpyright actx typing

Follow-up to the AppliedUndef chain-rule fix:
- Rename uppercase symbolic placeholder R to r_placeholder so PEP 8
  naming rules (N802/N803/N806) are satisfied; rename helper functions
  similarly.
- Drop redundant 0 start in range(0, len(terms)-1) (PIE808).
- Annotate actx as PyOpenCLArrayContext in test_recurrence_qbx.py so
  basedpyright can resolve .context and .queue attribute accesses
  (resolves three reportAttributeAccessIssue errors at lines 78/91/96).
  The now-redundant reportArgumentType ignores on lines 91/96 are
  removed.

hirish99 committed:
test_recurrence_qbx: scope pyright ignores to the actual failing rule

The previous attempt (annotating actx as PyOpenCLArrayContext) traded one
set of basedpyright errors for another: CI's view of actx_factory() is
ArrayContext (reportAssignmentType on the annotation), and once actx
becomes PyOpenCLArrayContext, actx.queue resolves to CommandQueue and
fails the next-arg type check. Revert to the original untyped pattern and
suppress reportAttributeAccessIssue on the three attribute accesses
(actx.context once, actx.queue twice) that don't resolve on the base
ArrayContext type CI sees.

hirish99 committed:
recurrence: silence SymPy-untyped basedpyright warnings; revert PIE808

Drives `basedpyright sumpy/recurrence.py` to 0 warnings:

- Annotate the two closure parameters (eval_r_placeholder_eps_derivs) as
  Expr -> Expr; kills the reportUnknownParameterType /
  reportMissingParameterType cluster.

- Per-line `# pyright: ignore[...]` on the SymPy-method-call lines that
  basedpyright cannot type without proper sympy stubs: .atoms, .diff,
  .xreplace, .subs, .doit, plus the *var unpack inside
  sp.Function("R")(*var, eps). Matches the file's existing style
  (pde_to_ode_in_r already uses ignore[reportAssignmentType] for the same
  reason).

- Revert the PIE808 fix on the true_recurrence sum-generator
  (range(0, len(terms)-1)) and add `# noqa: PIE808`. The PIE808 change
  was cosmetic and triggered a Generator[Unknown] warning under
  basedpyright that wouldn't otherwise have fired; keeping the original
  form sidesteps it and stays ruff-clean.

hirish99 committed:
Drop QBX integration (sumpy.recurrence_qbx) from this PR

The QBX-specific code wasn't going to be consumed downstream and was the
sole cause of two remaining CI failures:

- Conda Pytest's "ModuleNotFoundError: No module named 'meshmode'" came
  from test_recurrence_qbx.py's meshmode/pytential imports, which aren't
  in the conda test env.
- basedpyright's exit-1 came from 43 pre-existing reportUnknown* warnings
  in test_recurrence_qbx.py that the project's executionEnvironments
  config doesn't downgrade (root="test" matches a non-existent
  top-level test/ rather than sumpy/test/).

Removing recurrence_qbx.py + test_recurrence_qbx.py scopes the PR to the
parametric-recurrence math contribution alone, which the (now passing) 8
tests in test_recurrence.py exercise end-to-end against analytic Green's
functions for 2D/3D Laplace and 2D Helmholtz.

Also drops the doc/expansion.rst automodule reference for the removed
module and a stale comment in test_recurrence.py.

hirish99 commented:
@inducer All tests passing. Ready for merge.

hirish99 committed:
recurrence: use symbolic doit wrapper
